frontend/src/lib/keybind.ts

```diff
@@ -0,0 +1,246 @@
+   */
+  constructor(bindings: KeybindCallback[]) {
+    this.bindings = bindings;
+    this.bindingEntries = this.bindings.map((x) => x.key);
+  }
+
+  /**
+   * Attempt to convert a string to {@link Keys} enum
+   *
+   * @param str - The string to attempt to convert
+   * @returns undefined if unable to convert, otherwise {@link Keys}
+   */
+  private keyEnumFromString(str: string): Keys | undefined {
+    for (const [_, key] of Object.entries(Keys)) {
+      if (key == str) return key;
+    }
+
+    return undefined;
+  }
+
+  /**
+   * Handler for `keydown` events
+   *
+   * @remarks
+   * Requires a wrapper function to pass in `useKeyboardShortcuts` boolean store.
+   *
+   * Will run the given callback if all requirements are met from an item in the
+   * previously given {@link KeybindCallback} array.
+   *
+   * @param e - The `keydown` event of `KeyboardEvent` type
+   * @param useKeyboardShortcuts - A boolean of the user's `useKeyboardShortcuts` preference
+   */
+  onKeyDown(e: KeyboardEvent, useKeyboardShortcuts: boolean) {
+    if (!useKeyboardShortcuts || e.repeat) return;
+
+    const enum_key = this.keyEnumFromString(e.key);
+
+    if (!(enum_key && this.bindingEntries.includes(enum_key))) return;
+
+    // Conditionals to be able to later compare to a potentially undefined binding.shift/.al/.ctrl
+    const shift = e.shiftKey ? true : undefined;
+    const ctrl = e.ctrlKey ? true : undefined;
+    const alt = e.altKey ? true : undefined;
+
+    const binding = this.bindings.find((binding) => binding.key === enum_key);
+    if (binding && binding.alt == alt && binding.shift == shift && binding.ctrl == ctrl) {
+      e.preventDefault();
+      binding.callback(e);
+    }
+  }
+}
+
+/*
+Search callbacks
+*/
+
+/**
+ * Used to indicate what direction {@link navigateResults} should go (up/down)
+ */
+type Direction = 1 | -1;
+
+/**
+ * Utilized to bring the next/previous result into focus
+ *
+ * @remarks
+ * Utilizes the element that is currently in focus and the given
+ * direction to determine where and how to traverse the array of
+ * results. Moves a single result per call.
+ *
+ * @param direction - How to maneuver through the result list (up/down), indicated by {@link Direction}
+ */
+const navigateResults = (direction: Direction) => {
+  const currentActive = document.activeElement;
+
+  const results = [...document.getElementsByClassName('result')] as HTMLElement[];
+  let currentResultIndex: number | null = null;
+  let newIndex = 0;
+
+  results.forEach((el, index) => {
+    if (el.contains(currentActive)) currentResultIndex = index;
+  });
+
+  if (currentResultIndex != null) {
+    newIndex = currentResultIndex + direction;
+    if (0 > newIndex || newIndex >= results.length) {
+      newIndex = currentResultIndex;
+    }
+  }
+
+  const nextResult = results[newIndex];
+
+  // Get the anchor element and focus it
+  (nextResult.getElementsByClassName('result-main-link')[0] as HTMLElement)?.focus();
+};
```

Comment:
If we make a change to the class names in result, this would break without getting any type errors during compilation. We should therefore avoid all query selectors like `.getElementsByClassName` and instead give references to the relevant components as parameters to the functions. I would propose something like this
```ts
const navigateResults = (direction: Direction, refs: Refs) => {
  const currentActive = document.activeElement;
  const results = refs.results;

  let newIndex = 0;
  const currentResultIndex = results.findIndex((el) => el.getMainDiv().contains(currentActive));

  if (currentResultIndex > -1) {
    newIndex =
      currentResultIndex +
      match(direction)
        .with('up', () => -1)
        .with('down', () => 1)
        .exhaustive();

    if (newIndex < 0 || newIndex >= results.length) {
      newIndex = currentResultIndex;
    }
  }

  results[newIndex].focusMainLink();
};
```

Where `getMainDiv()` is a method defined in `Result.svelte`
```ts
  let mainDiv: HTMLDivElement;
  export const getMainDiv = () => mainDiv;
  ```
  with a bind like this
  ```html
  <div class="result flex min-w-0 grow flex-col space-y-0.5" bind:this={mainDiv}>
...
  ```
  and `Refs` is a new type in `keybind.ts` that holds the references
  ```ts
  export type Refs = {
  results: Result[];
};
```

 Structuring it this way allows us to see directly in `Result.svelte` that we use the reference to the div in other parts of the code, and should make the appropiate changes in all parts that use `.getMainDiv()` if we make changes to `Result.svelte`. It also allows us to move the link opening logic directly into `ResultLink.svelte` which can then be re-used in other parts of the code if we need to at some point in the future
 ```ts
   export const focus = () => {
    link.focus();
  };
  export const open = () => {
    link.click();
  };
  export const openCurrentTab = () => {
    window.location.href = link.href;
  };
  export const openNewTab = () => {
    window.open(link.href, '_blank', 'noopener');
  };
  export const hasFocus = () => document.activeElement == link
  ```

Reply:
Why `Refs` as it's own type rather than just replacing that with functions that take `Result[]` as the type? ie. instead of 
```ts
const navigateResults = (direction: Direction, refs: Refs) => {...}
```
would it be less complex to do
```ts
const navigateResults = (direction: Direction, result: Result[]) => {...}
```

Reply:
or so more could be tossed into `Refs`? like along side of results we could have the misspell link and the likes?

Reply:
> or so more could be tossed into Refs?

Exactly. The idea was to bundle the references together to reduce the number of arguments needed for the methods a bit